File: impulse-rust/examples/01_snapshot_basics.rs
//! Impulse Graph Engine — Example 01: Snapshot Creation & Zero-Copy Reading (Rust)
//!
//! Demonstrates:
//! 1. Programmatic creation of an immutable binary snapshot (.imps) using SnapshotWriter.
//! 2. Opening the snapshot via zero-copy OS memory mapping with SnapshotReader.
//! 3. Inspecting domains, relations, degree arrays, and point reachability.

use impulse_graph::{KeyType, SnapshotReader, SnapshotWriter};
use std::fs;
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("===============================================================");
    println!(" Impulse Graph Engine — Example 01: Snapshot Basics (Rust)");
    println!("===============================================================\n");

    let snapshot_path = "sample_basics.imps";

    // ------------------------------------------------------------------------
    // Step 1: Programmatic Snapshot Creation with SnapshotWriter
    // ------------------------------------------------------------------------
    println!("1. Creating binary snapshot file: {}...", snapshot_path);

    let mut writer = SnapshotWriter::new(snapshot_path);
    writer.add_domain(0, KeyType::Int64, "User");

    // CSR Topology: 4 Users (0, 1, 2, 3)
    // Node 0 -> [1, 2]
    // Node 1 -> [2, 3]
    // Node 2 -> [3]
    // Node 3 -> []
    let row_offsets = vec![0u32, 2, 4, 5, 5];
    let col_indices = vec![1u32, 2, 2, 3, 3];

    writer.add_relation(
        /*src_domain_id=*/ 0,
        /*tgt_domain_id=*/ 0,
        /*node_count=*/ 4,
        /*edge_count=*/ 5,
        row_offsets,
        col_indices,
    );

    writer.finalize()?;
    let file_size = fs::metadata(snapshot_path)?.len();
    println!("   -> Successfully wrote snapshot ({} bytes).\n", file_size);

    // ------------------------------------------------------------------------
    // Step 2: Zero-Copy Memory-Mapped Reading with SnapshotReader
    // ------------------------------------------------------------------------
    println!("2. Opening snapshot via zero-copy memory mapping...");
    let reader = SnapshotReader::open(snapshot_path)?;

    println!("   -> Magic:     0x{:X} ('IMPS')", reader.header().magic());
    println!("   -> Version:   {}", reader.header().version());
    println!("   -> Domains:   {}", reader.domain_count());
    println!("   -> Relations: {}\n", reader.relation_count());

    // ------------------------------------------------------------------------
    // Step 3: Inspect Relation Directory & Slices
    // ------------------------------------------------------------------------
    println!("3. Inspecting relation 0 topology:");
    if let Some(rel) = reader.get_relation_entry(0) {
        println!("   -> Node Count: {}", rel.node_count);
        println!("   -> Edge Count: {}", rel.edge_count);
    }

    let offsets = reader.get_row_offsets(0)?;
    let targets = reader.get_col_indices(0)?;

    for node in 0..4 {
        let start = offsets[node] as usize;
        let end = offsets[node + 1] as usize;
        let node_targets = &targets[start..end];
        println!(
            "   -> Node {} out-degree: {} edges (neighbors: {:?})",
            node,
            node_targets.len(),
            node_targets
        );
    }

    println!("\n4. Direct Point Reachability Queries:");
    println!("   -> Node 0 -> Node 1 reachable? {}", reader.is_reachable(0, 0, 1));
    println!("   -> Node 0 -> Node 3 reachable? {}", reader.is_reachable(0, 0, 3));

    // Cleanup
    if Path::new(snapshot_path).exists() {
        fs::remove_file(snapshot_path)?;
    }

    println!("\n[SUCCESS] Example 01 completed cleanly.");
    Ok(())
}

File: impulse-node/examples/01_snapshot_basics.js
/**
 * Impulse Graph Engine — Example 01: Snapshot Basics (Node.js)
 *
 * Demonstrates:
 * 1. Programmatic creation of an immutable binary snapshot (.imps) using Writer.
 * 2. Opening the snapshot via zero-copy memory mapping with Snapshot.
 * 3. Inspecting domains, relations, degree arrays, and point reachability.
 */

const fs = require('fs');
const { Writer, Snapshot } = require('../index.js');

function main() {
  console.log('===============================================================');
  console.log(' Impulse Graph Engine — Example 01: Snapshot Basics (Node.js)');
  console.log('===============================================================\n');

  const snapshotPath = 'sample_basics.imps';

  // --------------------------------------------------------------------------
  // Step 1: Programmatic Snapshot Creation with Writer
  // --------------------------------------------------------------------------
  console.log(`1. Creating binary snapshot file: ${snapshotPath}...`);

  const writer = new Writer(snapshotPath);
  writer.addDomain(0, 4, 'User'); // Key Type: INT64

  // CSR Topology: 4 Users (0, 1, 2, 3)
  // Node 0 -> [1, 2]
  // Node 1 -> [2, 3]
  // Node 2 -> [3]
  // Node 3 -> []
  const rowOffsets = new Uint32Array([0, 2, 4, 5, 5]);
  const colIndices = new Uint32Array([1, 2, 2, 3, 3]);

  writer.addRelation(
    0, // srcDomain
    0, // tgtDomain
    0, // encoding (CSR)
    4, // nodeCount
    5, // edgeCount
    0, // sectionFeatures
    rowOffsets,
    colIndices
  );

  writer.finalize();
  const fileSize = fs.statSync(snapshotPath).size;
  console.log(`   -> Successfully wrote snapshot (${fileSize} bytes).\n`);

  // --------------------------------------------------------------------------
  // Step 2: Zero-Copy Memory-Mapped Reading with Snapshot
  // --------------------------------------------------------------------------
  console.log('2. Opening snapshot via zero-copy OS memory mapping...');
  const snap = new Snapshot(snapshotPath);

  // --------------------------------------------------------------------------
  // Step 3: Point Reachability Queries
  // --------------------------------------------------------------------------
  console.log('\n3. Direct Point Reachability Queries:');
  console.log(`   -> Node 0 -> Node 1 reachable? ${snap.isReachable(0, 0, 0, 1) ? 'YES' : 'NO'}`);
  console.log(
    `   -> Node 0 -> Node 3 reachable? ${snap.isReachable(0, 0, 0, 3) ? 'YES (direct)' : 'NO (multi-hop path)'}`
  );

  // Cleanup
  snap.close();
  if (fs.existsSync(snapshotPath)) {
    fs.unlinkSync(snapshotPath);
  }

  console.log('\n[SUCCESS] Example 01 completed cleanly.');
}

main();
